analyses/cms-open-data-ttbar/jetassignment_training.py

```diff
@@ -0,0 +1,1347 @@
+                                jets[perms[...,2]]).mass).to_numpy()
+    
+    # combined pT of W and top_hadron
+    features[:,7] = ak.flatten((jets[perms[...,0]] + jets[perms[...,1]] + 
+                                jets[perms[...,2]]).pt).to_numpy()
+
+
+    # pt of every jet
+    features[:,8] = ak.flatten(jets[perms[...,0]].pt).to_numpy()
+    features[:,9] = ak.flatten(jets[perms[...,1]].pt).to_numpy()
+    features[:,10] = ak.flatten(jets[perms[...,2]].pt).to_numpy()
+    features[:,11] = ak.flatten(jets[perms[...,3]].pt).to_numpy()
+    
+    # btagCSVV2 of every jet
+    features[:,12] = ak.flatten(jets[perms[...,0]].btagCSVV2).to_numpy()
+    features[:,13] = ak.flatten(jets[perms[...,1]].btagCSVV2).to_numpy()
+    features[:,14] = ak.flatten(jets[perms[...,2]].btagCSVV2).to_numpy()
+    features[:,15] = ak.flatten(jets[perms[...,3]].btagCSVV2).to_numpy()
+    
+    # qgl of every jet
+    features[:,16] = ak.flatten(jets[perms[...,0]].qgl).to_numpy()
+    features[:,17] = ak.flatten(jets[perms[...,1]].qgl).to_numpy()
+    features[:,18] = ak.flatten(jets[perms[...,2]].qgl).to_numpy()
+    features[:,19] = ak.flatten(jets[perms[...,3]].qgl).to_numpy()
+    
+    #### calculate combination-level labels ####
+    permutation_labels = np.array(labels_dict[4])
+    
+    # which combination does the truth label correspond to?
+    which_combination = np.zeros(len(jets), dtype=int)
+    # no correct matches
+    which_anti_combination = np.zeros(labels.shape[0], dtype=int)
+    for i in range(12):
+        which_combination[(labels==permutation_labels[i,:]).all(1)] = i
+        which_anti_combination[np.invert((labels==permutation_labels[i,:]).any(1))] = i
+
+    # convert to combination-level truth label (-1, 0 or 1)
+    which_combination = list(zip(range(len(jets),), which_combination))
+    which_anti_combination = list(zip(range(labels.shape[0],), which_anti_combination))
+    
+    truth_labels = -1*np.ones((len(jets),12))
+    for i,tpl in enumerate(which_combination):
+        truth_labels[tpl]=1
+    for i,tpl in enumerate(which_anti_combination):
+        truth_labels[tpl]=0
+        
+        
+    #### flatten to combinations (easy to unflatten since each event always has 12 combinations) ####
+    labels = truth_labels.reshape((truth_labels.shape[0]*truth_labels.shape[1],1))
+    
+    return features, labels, which_combination
+
+
+# %% [markdown]
+# ### Defining a `coffea` Processor
+#
+# The processor returns the training features and labels we will use in our BDT
+
+# %% tags=[]
+# function to create column accumulator from list
+def col_accumulator(a):
+    return processor.column_accumulator(np.array(a))
+
+processor_base = processor.ProcessorABC
+class JetClassifier(processor_base):
+    def __init__(self, permutations_dict, labels_dict):
+        super().__init__()
+        self.permutations_dict = permutations_dict
+        self.labels_dict = labels_dict
+    
+    def process(self, events):
```

Comment:
Perhaps we do not have much of an alternative for now, but duplicating the processor code (and a few other piecse) from the main notebook is going to make maintenance difficult. Do you think that conceptually it would be possible to avoid all duplication and import common pieces from a common dependency? We do not need to do that here, but it would be interesting to know if that might in principle work (and we can refactor later on).

Reply:
I can probably remove the trijet combination calculation from this notebook. I just left it in so that we could compare the two methods for top mass reconstruction. Do you think it is best if this is removed then?

Reply:
I think that's the only part of the processor code which is really duplicated...

Reply:
Some of the configuration options at the top of the notebook also are, the metadata handling below for normalization is the same, and some cuts are as well. They aren't huge overlaps but I think what we need to eventually move towards is a setup where all this functionality is packaged up in common utility functions that are called from any processor we use for this analysis.

I would say leave the trijet combination in here for now and in the longer term we can try to find a solution that avoids overlaps.

Reply:
Yeah.. I think without dependency management for dask workers, handling this in a great way is not possible at the moment

Reply:
Could you please open an issue noting down these things here so we don't forget about them and can follow up on them in the future? We can keep this PR here the way it is.

Reply:
#129